AGENTS.md

```diff
@@ -0,0 +1,663 @@
+
+Downloaded via `hack/fetch_ext_bins.sh`:
+
+- `kube-apiserver`
+- `etcd`
+- `kubectl`
+
+Location: `/tmp/kubebuilder/bin/`
+
+---
+
+## Integration Points
+
+### With CAPM3
+
+CAPM3's Metal3DataTemplate references IPPool:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3DataTemplate
+spec:
+  networkData:
+    networks:
+      ipv4:
+        - id: provisioning
+          link: eth0
+          ipAddressFromIPPool: pool1
+```
+
+### With Cluster API
+
+- Supports CAPI v1beta1/v1beta2 IPAddressClaim resources
+- Compatible with CAPI's InfraCluster IPAM contract
+
+### E2E Testing with CAPM3
+
+**Important:** IPAM does not have standalone e2e tests. All e2e testing is
+performed through the CAPM3 repository, which serves as the central e2e
+testing hub for the entire Metal3 stack (BMO, CAPM3, IPAM).
+
+**How e2e works for IPAM:**
+
+1. CAPM3 e2e tests provision real bare metal clusters using Metal3
+2. Metal3DataTemplate resources reference IPPool for network configuration
+3. IPAM allocates IPs during cluster provisioning
+4. Tests verify IP allocation in Metal3Data and BareMetalHost network-data
+
+**IPAM's role in e2e:**
+
+- Provides IP addresses for provisioning, BMC, and external networks
+- IPPool resources are created as part of test cluster setup
+- IPClaims are automatically created by CAPM3 controllers
+- Allocation failures cause cluster provisioning to fail
+
+**When cutting new release branches:**
+
+When creating a new IPAM release branch (e.g., `release-1.11`), corresponding
+updates are needed in CAPM3:
+
+1. CAPM3's e2e tests reference IPAM container images
+2. New IPAM branches need matching Prow job configurations in project-infra
+3. CAPM3's `metadata.yaml` and clusterctl configuration may need updates
+4. E2E test manifests in CAPM3 may reference specific IPAM versions
+
+**Running e2e tests:**
+
+```bash
+# E2E tests are triggered via Prow jobs on CAPM3 PRs
+# They can also be triggered on IPAM PRs using /test comments:
+/test metal3-centos-e2e-integration-test-main
+/test metal3-ubuntu-e2e-integration-test-main
+```
+
+See `project-infra/prow/config/jobs/metal3-io/ip-address-manager.yaml` for
+available e2e job triggers.
+
+---
+
+## Common Pitfalls
+
+1. **Pool Exhaustion** - Claim stays pending when no IPs available
+2. **Pre-allocation Conflicts** - IP outside pool ranges
+3. **Namespace Mismatch** - Cross-namespace refs not supported
+4. **Manual IPAddress Creation** - Always use IPClaim
+5. **Pool Updates** - Avoid modifying pools in use
+6. **Forgetting `make generate`** - After API changes
+
+---
+
+## AI Agent Behavioral Guidelines
+
+### Critical Rules
+
+1. **Use single commands** - Do not concatenate multiple commands with `&&` or
+   `;` to avoid interactive permission prompts from the user. Run one command
+   at a time.
+
+2. **Avoid excessive filtering** - Do not use `head`, `tail`, or excessive
+   `grep` piping unless specifically needed. Read full output first.
```

Comment:
This sounds potentially counter productive since the output can be too much and just fill the context window. I would skip it.

Reply:
It is double-edged for sure, I have a lot of frustration coming from AI doing | tail -10 and missing everything important, all the time. I can see it  is not productive for log reading etc, but for joining commands its usually only problematic to cut off the output.

I'll refine it to make it more of a compromise and try make agent use it reasonably.